### ISSUE TITLE
Polymorphism support issue

### ISSUE THREAD

**Issue by ykh1988** [domaindrivendev/Swashbuckle.AspNetCore#842]
I am using Swashbuckle v3.0.

I am not sure weather this is a bug or not, but polymorphism is not working as it should. I have the following classes:

```
BasePersonDocumentDto
{
   Id,
   Number
}

IdentityDto: BasePersonDocumentDto
{
 
}

PassportDto: BasePersonDocumentDto
{
 VisaNumber
}
```

To apply Inheritance & Polymorphism, i have created a schema and document filters. Below are the code i used.

```
    public class PolymorphismSchemaFilter<T> : ISchemaFilter
    {
        private List<Type> derivedTypes = new List<Type>() { typeof(IdentityDto), typeof(PassportDto) };

        public void Apply(Schema model, SchemaFilterContext context)
        {
            if (!derivedTypes.Contains(context.SystemType)) return;

            var baseSchema = new Schema() { Ref = "#/definitions/" + typeof(T).Name };
            var clonedBaseSchema = new Schema
            {
                Properties = model.Properties,
                Type = model.Type,
                Required = model.Required
            };

            model.AllOf = new List<Schema> { baseSchema, clonedBaseSchema };

            //Reset properties for they are included in allOf, should be null but code does not handle it
            model.Properties = new Dictionary<string, Schema>();
        }
    }
```
```
    public class PolymorphismDocumentFilter<T> : IDocumentFilter
    {
        private List<Type> derivedTypes = new List<Type>() { typeof(IdentityDto), typeof(PassportDto) };

        public void Apply(SwaggerDocument swaggerDoc, DocumentFilterContext context)
        {
            const string discriminatorName = "discriminator";

            var baseSchema = context.SchemaRegistry.Definitions[typeof(T).Name];

            //Discriminator property
            baseSchema.Discriminator = discriminatorName;
            baseSchema.Required = new List<string> { discriminatorName };

            if (!baseSchema.Properties.ContainsKey(discriminatorName))
                baseSchema.Properties.Add(discriminatorName, new Schema { Type = "string" });

            //Register dervied classes
            foreach (var item in derivedTypes)
                context.SchemaRegistry.GetOrRegister(item);
        }
    }
```
The generated definitions are:

```
//This is for BasePersonDocumentDto and for IdentityDto
"BasePersonDocumentDto":{  
         "required":[  
            "discriminator"
         ],
         "type":"object",
         "properties":{  
            "number":{  
               "type":"string"
            },
            "id":{  
               "format":"int32",
               "type":"integer"
            },
            "discriminator":{  
               "type":"string"
            }
         },
         "discriminator":"discriminator"
      }

"VisaDto":{  
         "required":[  
            "discriminator"
         ],
         "type":"object",
         "properties":{  
            "number":{  
               "type":"string"
            },
            "id":{  
               "format":"int32",
               "type":"integer"
            },
            "visaNumber":{  
               "type":"string"
            },
            "discriminator":{  
               "type":"string"
            }
         },
         "discriminator":"discriminator"
      }

```
Now to test, i create an InputDto that have a List<BasePersonDocumentDto> property. I tried passing the following JSON, but both passed items are shown as BasePersonDocumentDto even though the discriminator is passed correctly. Here is the JSON i passed:

```
{  
   "documentsDto":[  
      {  
         "discriminator":"IdentityDto"
      },
      {  
         "visaNumber":"RRXX323TR",
         "discriminator":"PassportDto"
      }
   ]
}
```

**Comment by cleftheris** [comment]
Hi you are reporting that you see the items in your model server side deserialized as the base class `BasePersonDocumentDto`. Am I right?

I think this issue has nothing todo with swashbuckler. You should be looking into custom model binding. [See an example here](https://www.tutorialdocs.com/article/webapi-data-binding.html)

**Comment by lemoinem** [comment]
With #1041, we now have basic support, but without a way to tell Swashbuckle what property is the discriminator, nor what is the discriminatorMap, this is somewhat useless.

Sure, it can easily be generated using a custom SchemaFilter, but wouldn't it be easier to have an annotation on the property take care of that?

It could be as easy as having:
* `[SwaggerDiscriminator]` (no map)
* `[SwaggerDiscriminator(new Dictionnary<PropertyType, Type>{ {PropertyType.A, typeof(A)}, ... })]` (enum-based map)

What do you think?

**Comment by lemoinem** [comment]
I forgot that dictionary cannot be used as Attribute parameters (doh!!!), so we would need a `[SwaggerDiscriminatorEntry(PropertyType.A, typeof(A))]` as well...

**Comment by vbjay** [comment]
Since I was dealing with these issues with a .net core 3.1 api, let me point you to a few things that help

- https://stackoverflow.com/a/60549017/815813 with a minor change of 
```c#
Reference = new OpenApiReference() { ExternalResource = "#/components/schemas/" + typeof(T).Name }
```
instead of #/definitions

- https://stackoverflow.com/a/59789176/815813 allows you to opt in by using a JsonConvertAttribute where needed to serialize with derived class properties.

**Comment by domaindrivendev** [comment]
I believe this should be resolved with the enhanced polymorphism/inheritance support available as of `v5.6.0`:

https://github.com/domaindrivendev/Swashbuckle.AspNetCore#inheritance-and-polymorphism
